File: Task/Accumulator-factory/Objective-C/accumulator-factory.m
#import <Foundation/Foundation.h>

typedef double (^Accumulator)(double);

Accumulator accumulator_factory(double initial) {
    __block double sum = initial;
    Accumulator acc = ^(double n){
        return sum += n;
    };
    return [[acc copy] autorelease];
}

int main (int argc, const char * argv[]) {
    NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];

    Accumulator x = accumulator_factory(1);
    x(5);
    accumulator_factory(3);
    NSLog(@"%f", x(2.3));
	
    [pool drain];
    return 0;
}

File: Task/Accumulator-factory/Zen-C/accumulator-factory.c
fn accumulator<T>(acc: T) -> fn(T) -> T {
    return fn(f: T) -> T {
        acc += f;
        return acc;
    }
}

fn main() {
    // Example with f64s.
    let x = accumulator(1.0);
    x(5.0);
    accumulator(3.0);
    println "{x(2.3):g}";

    // Example with ints.
    let y = accumulator(1);
    y(5);
    accumulator(3);
    println "{y(2)}";
}
